core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -781,6 +781,23 @@ private[spark] class TaskSetManager(
     sortedTaskSetQueue
   }
 
+  /**
+   * Called by TaskScheduler when an executor is lost, but the reason is not yet known. This method
+   * does not fail any tasks related to the executor. Instead, tasks are left as is, but the
+   * executor is removed from the list of live executors, so no new tasks are scheduled. Pending
+   * tasks for the executor are re-queued.
+   */
+  override def disableExecutor(execId: String, host: String): Unit = {
+    for (index <- getPendingTasksForExecutor(execId)) {
+      addPendingTask(index, readding = true)
+    }
+    for (index <- getPendingTasksForHost(host)) {
+      addPendingTask(index, readding = true)
+    }
+    // recalculate valid locality levels and waits when executor is disabled.
+    recomputeLocality()
```

Comment:
It seems to not be expensive; looks `O(1)` except in rare cases where most executors with pending tasks are dead, in which case it would be `O(number executors)`.